plugins/image-prioritizer/class-image-prioritizer-img-tag-visitor.php

```diff
@@ -28,7 +42,37 @@ final class Image_Prioritizer_Img_Tag_Visitor extends Image_Prioritizer_Tag_Visi
 	 */
 	public function __invoke( OD_Tag_Visitor_Context $context ): bool {
 		$processor = $context->processor;
-		if ( 'IMG' !== $processor->get_tag() ) {
+		$tag       = $processor->get_tag();
+
+		// Handle opening and closing `<picture>` tags.
+		if ( 'PICTURE' === $tag ) {
```

Comment:
Instead of relying on this `__invoke` function being called on the `PICTURE` and then also on the child `SOURCE` tags (and `IMG`) tag, I suggest instead that when `'PICTURE' === $tag` that this here call a new method like `process_picture()` and that this method would:

1. Call `$processor->set_bookmark( 'img-prioritizer-picture' )`
2. Add a `while( $processor->next_tag()` loop. Note that our subclass's `next_tag()` method [always visits tag closers](https://github.com/WordPress/performance/blob/69eca9a7936701475f84a0cabbfec36e1cb3fe27/plugins/optimization-detective/class-od-html-tag-processor.php#L214-L219) until `$processor->get_tag() === 'picture' && $processor->is_tag_closer()`.
3. Gather up the attributes data from the `SOURCE` tags and the final `IMG` tag.
4. Finally reset the processor back to where you started: `$processor->set_bookmark('img-prioritizer-picture')`.
5. Process the data collected from the child tags, and call `$context->link_collection->add_link()`, leaving the other location as-is for just for processing bare `IMG` tags.

This approach of a sub-loop was also taken by Embed Optimizer in [`embed_optimizer_update_markup()`](https://github.com/WordPress/performance/blob/69eca9a7936701475f84a0cabbfec36e1cb3fe27/plugins/embed-optimizer/hooks.php#L179-L308).